crates/core/src/delta_datafusion/table_provider.rs

```diff
@@ -662,24 +702,58 @@ impl<'a> DeltaScanBuilder<'a> {
         let file_source =
             file_source.with_schema_adapter_factory(Arc::new(DeltaSchemaAdapterFactory {}))?;
 
-        let file_scan_config =
-            FileScanConfigBuilder::new(self.log_store.object_store_url(), file_schema, file_source)
-                .with_file_groups(
-                    // If all files were filtered out, we still need to emit at least one partition to
-                    // pass datafusion sanity checks.
-                    //
-                    // See https://github.com/apache/datafusion/issues/11322
-                    if file_groups.is_empty() {
-                        vec![FileGroup::from(vec![])]
-                    } else {
-                        file_groups.into_values().map(FileGroup::from).collect()
-                    },
-                )
-                .with_statistics(stats)
-                .with_projection_indices(self.projection.cloned())
-                .with_limit(self.limit)
-                .with_table_partition_cols(table_partition_cols)
-                .build();
+        // If using local filesystem, register a root-scoped store and scan against that
+        let scan_store_url = if self.log_store.config().location.scheme() == "file" {
+            if allow_unrestricted_file_access {
+                let url = ObjectStoreUrl::parse("delta-rs://file-root").unwrap();
+                // Ensure a root store is registered for the session runtime
+                self.session
+                    .runtime_env()
+                    .register_object_store(url.as_ref(), self.log_store.root_object_store(None));
+                url
+            } else {
+                // By default, do NOT register a root object store for the file scheme.
+                // Only files within the table directory will be accessible.
+                self.log_store.object_store_url()
+            }
+        } else {
+            if need_bucket_root_store {
+                // Build a root-scoped store identifier unique per (scheme, host)
+                let loc = &self.log_store.config().location;
+                let host = loc.host_str().unwrap_or("-");
+                // TODO. May be a security risk to allow absolute paths outside the bucket/account.
+                // When need_bucket_root_store is true, the code registers a bucket-root object
+                // store that could access ANY object in the bucket, not just those in the
+                // table directory. This has similar security implications to the file-root store above.
```

Comment:
Note. I have not added an environment variable to allow this like I did for filesystems. I'm not as sure this is necessary since object stores have more detailed bucket permissions. But, I guess it could still be a problem for delta-rs running as a superuser with malicious delta tables.